### PR TITLE
Switch to schema driven templates for generation and UI

### PR DESCRIPTION
**Pull request by chrimc62**
In order to make built-in types and schemas easier to consume switch to a schema file for every type.  This makes the connection between types and schemas clearer and also provides a place to provide information to Composer on extra schema to control generation.

Switched template generation to always be over locale, property, entity.  This works because of the file caching we have, i.e. if something depends on nothing like build.cmd it will get built once.  If it depends on property, once for all entities.  This also enables using the same template for multiple entities.

Made runbot/testbot pick-up latest SDK bits by default.

### CODE REVIEW

File: experimental/generation/generator/packages/library/templates/generator.lg [hunk 1]
```diff
@@ -74,275 +74,252 @@
...
+        "generic-change.dialog",
+        "generic-help.dialog"
+    ]}
+
+# dateFormatTemplates()
+- ${[
+        "genericEntity.lu",
+        "genericEntity.lg",
+        "genericProperty.lg",
+        "dateFormat-add.dialog",
```
Comment (coldplaying42) [inline experimental/generation/generator/packages/library/templates/generator.lg:105]: Could we incorporate dateFormat-add.dialog into genericEntity-add.dialog as the specific case?

File: experimental/generation/generator/packages/library/templates/generator.lg [hunk 2]
```diff
@@ -74,275 +74,252 @@
...
-                        ${setOrPush()}
-                    ]
-                }
-            ]
-        }
-    ]               
-    ${designerOpEntity('Add', entity)}     
-}
-```
+            "elseActions":[${else}]
```
Comment (coldplaying42) [inline experimental/generation/generator/packages/library/templates/generator.lg:234]: I am not sure if drafting the dialog in this way, could minimum and maximum appear in the same dialog ?

File: experimental/generation/generator/packages/library/templates/generator.lg [hunk 3]
```diff
@@ -558,18 +542,30 @@ ${if(length(entities[entity]) > 1, `    - @ ${entity} ${entity}`, '')}```
...
+> Entities that have units
+- ${contains(['age', 'dimension', 'money', 'temperature'], entity)}
+
+# units
+> Property units if specified
+- ${propertySchema.$units}
+
+# isLUISEntity(entity)
+> Assume LUIS entity unless built-in utterance
+- ${entity != 'utterance'}
```
Comment (coldplaying42) [inline experimental/generation/generator/packages/library/templates/generator.lg:565]: I think BreadEntity is also not built-in LUIS entity

File: experimental/generation/generator/packages/library/templates/standard/array.json [hunk 4]
```diff
@@ -0,0 +1,10 @@
+{
+    "array": {
+        "$comment": "Support arrays.",
+        "uniqueItems": {
+            "type": "boolean",
+            "title": "Unique Values",
+            "description": "Require all values to be unique."
+        }
+    }
+}
```
Comment (coldplaying42) [inline experimental/generation/generator/packages/library/templates/standard/array.json:10]: should this be a json file?

File: experimental/generation/generator/packages/library/templates/standard/duration.tschema [hunk 5]
```diff
@@ -0,0 +1,9 @@
+{
+    "$schema": "http://json-schema.org/draft-07/schema",
+    "type": "string",
+    "format": "duration",
+    "$generator": {
+        "$comment": "TODO: We need to convert to P[nY][nM][nD][T[nH][nM][nS]] https://tools.ietf.org/html/rfc3339#appendix-A",
+        "$ref": "array.json"
+    }
+}
```
Comment (coldplaying42) [inline experimental/generation/generator/packages/library/templates/standard/duration.tschema:9]: The file name should be duration.schema

File: experimental/generation/generator/packages/library/templates/standard/en-us/phonenumberProperty.en-us.lg.lg [hunk 6]
```diff
@@ -0,0 +1,7 @@
+[standard.en-us.lg](standard.en-us.lg)
+
+# filename
+- ${propertyFilename()}
+
+# template
+- ${genericMessages('phone number')}
```
Comment (coldplaying42) [inline experimental/generation/generator/packages/library/templates/standard/en-us/phonenumberProperty.en-us.lg.lg:7]: should be stringMessages ?

File: experimental/generation/generator/packages/library/templates/standard/en-us/uriProperty.en-us.lg.lg [hunk 7]
```diff
@@ -4,16 +4,4 @@
...
-                        unexpectedSetDefault(),
-                        unexpectedUpdateDefault(),
-                        removeDefault(),
-                        clearDefault())}'
-${helpTemplates(repromptDefault('URL'), helpDefault('URL'))}
-${showTemplate(showDefault())}
-${nameTemplate()}
-${valueTemplate()}
-```
+- ${genericMessages('URL')}
```
Comment (coldplaying42) [inline experimental/generation/generator/packages/library/templates/standard/en-us/uriProperty.en-us.lg.lg:7]: Could be stringMessages ?

File: experimental/generation/generator/packages/library/templates/standard/integer.lg [hunk 8]
```diff
@@ -1,8 +1,7 @@
 [generator.lg](../generator.lg)
 
 # entities
-- ${['number']}
+- ${['integer']}
```
Comment (coldplaying42) [inline experimental/generation/generator/packages/library/templates/standard/integer.lg:4]: integer is not built-in LUIS entity, does it work ?

File: experimental/generation/generator/packages/library/templates/standard/phonenumber.schema [hunk 9]
```diff
@@ -1,18 +1,8 @@
...
-        "phonenumberEntity.lu",
-        "phonenumberEntity.lg",
-        "phonenumberEntity-add-string.dialog",
-        "phonenumberEntity-remove-string.dialog",
-        "generic-change.dialog",
-        "generic-clear.dialog",
-        "generic-help.dialog",
-        "generic-show.dialog"
-    ]
+    "format": "phonenumber",
```
Comment (coldplaying42) [inline experimental/generation/generator/packages/library/templates/standard/phonenumber.schema:4]: I think there is no json format named phonenumber.

File: experimental/generation/generator/packages/library/templates/generator.lg (hunk 3 above)
Comment (chrimc62) [reply experimental/generation/generator/packages/library/templates/generator.lg:565]: But it is defined in LUIS as a custom entity.

---
In reply to: [570818149](https://github.com/microsoft/BotBuilder-Samples/pull/3061#discussion_r570818149) [](ancestors = 570818149)

File: experimental/generation/generator/packages/library/templates/generator.lg (hunk 1 above)
Comment (chrimc62) [reply experimental/generation/generator/packages/library/templates/generator.lg:105]: I'm going to leave this for now.

---
In reply to: [570805648](https://github.com/microsoft/BotBuilder-Samples/pull/3061#discussion_r570805648) [](ancestors = 570805648)

File: experimental/generation/generator/packages/library/templates/standard/phonenumber.lg [hunk 10]
```diff
@@ -0,0 +1,7 @@
+[generator.lg](../generator.lg)
+
+# entities
+- ${['phonenumber']}
```
Comment (coldplaying42) [inline experimental/generation/generator/packages/library/templates/standard/phonenumber.lg:4]: ${['phonenumber', 'utterance']}

**Review by coldplaying42 — APPROVED** [review]

File: experimental/generation/generator/packages/library/templates/standard/boolean.schema [hunk 11]
```diff
@@ -0,0 +1,4 @@
+{
+    "$schema": "http://json-schema.org/draft-07/schema",
+    "type": "boolean"
+}
```
[the remark this answers is not in the document]
Comment (chrimc62) [reply experimental/generation/generator/packages/library/templates/standard/boolean.schema:4]: I didn't think an array of booleans make sense and there are no other switches.

**Pull request merged**